### ISSUE TITLE
Recursive Lazy Sequence

### ISSUE THREAD

**Issue by tgriesser** [transduce/transduce#29]
I'm looking to be able to recursively be able to process 
values as lazy sequences

Let's take this code as a simple example:

```js
var list = ['a', 'b', 'c', ['d', 'e', 'f'], 'g']
var t = map((val) => `${val}1`)
var seq  = sequence(t, list)

seq.next() // a1
seq.next() // b1
seq.next() // c1
seq.next() // ['d', 'e', 'f']1
seq.next() // g1
```

What I want to accomplish is something like the following, where items can be joined into the lazy sequence on demand:

```js
var list = ['a', 'b', 'c', ['d', 'e', 'f'], 'g']

var t = compose(
  transduce((step, result, val) => {
    if (isArray(val)) return into(result, t, val) // would lazily "push" into the stepper
    return step(result, val)
  }),
  map((val) => `${val}1`)
)

var seq  = sequence(t, list)

seq.next() // a1
seq.next() // b1
seq.next() // c1
seq.next() // d1
seq.next() // e1
seq.next() // f1
seq.next() // g1
```

Any ideas on how it might be possible to achieve something similar to the above?

I'm pretty sure there's also a bug in the currently lazy sequence implementation. I've found it possible to get stuck in an infinite loop - it seems to not break when the list is exhausted. I didn't investigate too deeply but didn't have the same problem when using the code from transducers.js

One other thing I've noticed while trying to do things in the library currently, is that the code 
splitting and binding/applying of context has made things very difficult to follow. It's difficult to know
what the value of "this" is, and almost might be easier if stuff was done functionally in places where the new `transducer` function is being utilized.

**Comment by kevinbeaty** [comment]
> Any ideas on how it might be possible to achieve something similar to the above?

I'd have to think about that... seems like a modification of `cat`, but not sure

> I'm pretty sure there's also a bug in the currently lazy sequence implementation.

Bummer.  Filed #30

> It's difficult to know what the value of "this" is, and almost might be easier if stuff was done functionally in places where the new transducer function is being utilized.

I'm certainly open for suggestions... Filed #31 to discuss

**Comment by tgriesser** [comment]
> I'd have to think about that... seems like a modification of cat, but not sure

Yeah I'll think on it a little more as well - it would be super useful for a case I'm trying to address to sort of lazily flatten out an arbitrarily nested structure into a single stream of values based on a set of conditions.

Will comment on the others on the other tickets.

**Comment by tgriesser** [comment]
Ah, something like this sounds like it might be part of what I'm after:

http://clojure.github.io/clojure/clojure.core-api.html#clojure.core/lazy-cat
